geopandas/geodataframe.py

```diff
@@ -276,6 +285,11 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
                 raise ValueError("Unknown column %s" % col)
             except Exception:
                 raise
+            # specially check shape of level for consistency (shapely raises a
+            # TypeError but pygeos raises a ValueError)
+            if hasattr(level, "ndim") and level.ndim != 1:
+                raise ValueError("Must pass array with one dimension only.")
```

Comment:
Maybe use a similar message here as you added above? (about not supporting a column name that is duplicated) 
Since I think you also only get here when passing a column name to `set_geometry`

Reply:
The thing I was worried with in doing that is if you passed an `Index` or `MultiIndex` to set_geometry - not that I've ever tried to do that. Right now `Index` will fail anyway 
```python
df = GeoDataFrame({"geometry": [Point(0, 0), Point(1, 1)]})
df.set_geometry(pd.Index(["geometry"]))
Traceback (most recent call last):
  File "C:/Data/myDocuments/Code/geopandas_test/1738_keepna.py", line 40, in <module>
    df.set_geometry(pd.Index(["geometry"]))
  File "c:\data\mydocuments\code\geopandas_test\geopandas-m-richards\geopandas\geodataframe.py", line 312, in set_geometry
    level = _ensure_geometry(level, crs=crs)
  File "c:\data\mydocuments\code\geopandas_test\geopandas-m-richards\geopandas\geodataframe.py", line 46, in _ensure_geometry
    out = from_shapely(data, crs=crs)
  File "c:\data\mydocuments\code\geopandas_test\geopandas-m-richards\geopandas\array.py", line 151, in from_shapely
    return GeometryArray(vectorized.from_shapely(data), crs=crs)
  File "c:\data\mydocuments\code\geopandas_test\geopandas-m-richards\geopandas\array.py", line 280, in __init__
    raise ValueError(
ValueError: 'data' should be a 1-dimensional array of geometry objects.
``` 
as we end up with a (2,1) shaped array being passed to `from_shapely()`. This does change the error, so probably should be handled better.